### PR TITLE
For gff formats, lines starting with # will be passed

### LINKED ISSUE #898: Impossible to read GFF with comments or directives with GffFileIn (seqan 2.0.0.0)
Trying to read Gff with comment (a line begin with '#' char according to http://www.sequenceontology.org/gff3.shtml) or directives (begin with "##") throws an error : 
```
Unexpected character '\n' found.
```

You can use the Gff_IO demo code (gff_stream_read.cpp) : 
``` c++
#include <seqan/basic.h>
#include <seqan/gff_io.h>

using namespace seqan;

// USAGE: gff_stream_read IN.gff
//
// Read GFF file and print the positions (reference, position) to stdout.

int main(int argc, char ** argv)
{
    if (argc != 2)
    {
        std::cerr << "USAGE: " << argv[0] << " IN.gff\n";
        return 1;
    }

    try
    {
        GffFileIn gffIn(argv[1]);
        GffRecord record;
        while (!atEnd(gffIn))
        {
            readRecord(record, gffIn);

            // Note that we print the position 1-based since we use text output
            // whereas it is 0-based in the GffRecord.
            std::cout << record.ref << "\t" << (record.beginPos + 1) << "\n";
        }
    }
    catch (Exception & e)
    {
        std::cerr << e.what() << std::endl;
        return 1;
    }

    return 0;
}
```

with a GFF3 file with comment (or directive in this case) : 
```
##gff-version 3
ctg123	.	mRNA	1300	9000	.	+	.	ID=mrna0001;Name="sonichedgehog;hehe"
ctg123	.	exon	1300	1500	.	+	.	ID=exon00001;Parent=mrn a0001;Name
ctg123	.	exon	1050	1500	.	+	.	ID=exon00002;Name;Parent=mrna0001
```
I am not sure but I think the problem is in gff_io_base.h in function `readRecord`, at line 354  
``` c++
skipUntil(iter, NotFunctor<OrFunctor<EqualsChar<'#'>, IsWhitespace> >());  //skip commments and empty lines
```

In the case of a comment (or directive), it just skip '#' chars without going to the next line and so the next part of the function try to parse the comment as an GFF entry.

### PR DESCRIPTION
**Pull request by xenigmax**
This will close #898 . The lines staring with '#' will be passed.

See PR #948 for develop.

### CODE REVIEW

File: include/seqan/gff_io/gff_io_base.h [hunk 1]
```diff
@@ -349,9 +349,15 @@ void readRecord(GffRecord & record, CharString & buffer, TFwdIterator & iter)
 {
     IsNewline isNewline;
 
-    clear(record);
+    if (value(iter) == '#') // skip this line and find another record
+    {
+        skipLine(iter);
+        readRecord(record, buffer, iter);
```
Comment (weese) [inline include/seqan/gff_io/gff_io_base.h]: Please replace the recursion by a loop

Comment (xenigmax) [inline include/seqan/gff_io/gff_io_base.h]: @weese Thank you for the comment! Please feel free to let me know for other considerations.

File: include/seqan/gff_io/gff_io_base.h [hunk 2]
```diff
@@ -349,9 +349,12 @@ void readRecord(GffRecord & record, CharString & buffer, TFwdIterator & iter)
 {
     IsNewline isNewline;
 
-    clear(record);
+    // skip commented lines
+    while( value(iter) == '#' )
```
Comment (weese) [inline include/seqan/gff_io/gff_io_base.h]: Please mind the coding style and the possible EOF, i.e.

```
while (!atEnd(iter) && value(iter) == '#')
```

Comment (xenigmax) [inline include/seqan/gff_io/gff_io_base.h]: OK. I will be more aware of those things.
Thanks!

File: include/seqan/gff_io/gff_io_base.h [hunk 3]
```diff
@@ -349,9 +349,12 @@ void readRecord(GffRecord & record, CharString & buffer, TFwdIterator & iter)
 {
     IsNewline isNewline;
 
-    clear(record);
+    // skip commented lines
+    while( !atEnd(iter) && value(iter) == '#' )
```
Comment (weese) [inline include/seqan/gff_io/gff_io_base.h]: Now, it's only the spacing:

```
while (!atEnd(iter) && value(iter) == '#')
```

Comment (xenigmax) [inline include/seqan/gff_io/gff_io_base.h]: Thank you for the comments!

File: include/seqan/gff_io/gff_io_base.h [hunk 4]
```diff
@@ -349,9 +349,12 @@ void readRecord(GffRecord & record, CharString & buffer, TFwdIterator & iter)
 {
     IsNewline isNewline;
 
-    clear(record);
+    // skip commented lines
+    while(!atEnd(iter) && value(iter) == '#')
```
Comment (weese) [inline include/seqan/gff_io/gff_io_base.h]: Once more (missing space behind while):
```
while (!atEnd(iter) && value(iter) == '#')
```

Comment (weese) [inline include/seqan/gff_io/gff_io_base.h]: Please mind our C++ style guide (online on the readthedocs site) and consider to use uncrustify with our config (uncrustify.conf)

Comment (xenigmax) [inline include/seqan/gff_io/gff_io_base.h]: Oh.. I'm so sorry that you had to take your time for this simple change.
Perhaps your comments are longer than my modifications.

This is my first pull request to SeqAn, so I will be aware of coding conventions and others in the next time.
Thanks, again!

**Comment by esiragusa** [comment]
@weese @xenigmax please don't forget to make a PR to develop!

**Comment by esiragusa** [comment]
Now readRecord() handles comment-only lines but not end-of-line comments. It seems to me that the latter are allowed in Gtf2 but not in Gff3?
* Gtf2 spec (http://mblab.wustl.edu/GTF22.html) says: "Comments begin with a hash ('#') and continue to the end of the line. Nothing beyond a hash will be parsed. These may occur anywhere in the file, including at the end of a feature line."
* Gff3 spec (http://www.sequenceontology.org/gff3.shtml) says: "End-of-line comments (comments preceeded by # at the end of and on the same line as a feature or directive line) are not allowed."

Apart from that, the PR is okay for me. It would be good to add some comments line in test files e.g. https://github.com/seqan/seqan/blob/master/tests/gff_io/example.gff. However, when reading and writing such file, comments will not be written. So tests like https://github.com/seqan/seqan/blob/master/tests/gff_io/test_gff_io.h#L103 would fail .

**Comment by xenigmax** [comment]
I will make another pull request including Enrico's suggestion about test cases.

**Pull request closed without merging**